### PR TITLE
Part 9. Implement raw bidder response stage

### PR DESCRIPTION
**Pull request by legendko**

### CODE REVIEW

**Comment by 4lexvav** [comment]
[Implement BidRequestHook for ortb2blocking module](https://trello.com/c/sDQYxKl6/25-implement-bidrequesthook-for-ortb2blocking-module)

**Review by hhhjort — APPROVED** [review]
LGTM

File: exchange/exchange_test.go [hunk 1]
```diff
@@ -4466,7 +4479,7 @@ type validatingBidder struct {
 	mockResponses map[string]bidderResponse
 }
 
-func (b *validatingBidder) requestBid(ctx context.Context, bidderRequest BidderRequest, conversions currency.Conversions, reqInfo *adapters.ExtraRequestInfo, adsCertSigner adscert.Signer, bidRequestOptions bidRequestOptions, alternateBidderCodes openrtb_ext.ExtAlternateBidderCodes) (seatBids []*pbsOrtbSeatBid, errs []error) {
+func (b *validatingBidder) requestBid(ctx context.Context, bidderRequest BidderRequest, conversions currency.Conversions, reqInfo *adapters.ExtraRequestInfo, adsCertSigner adscert.Signer, bidRequestOptions bidRequestOptions, alternateBidderCodes openrtb_ext.ExtAlternateBidderCodes, executor hookexecution.StageExecutor) (seatBids []*pbsOrtbSeatBid, errs []error) {
```
Comment (bsardo) [inline exchange/exchange_test.go]: There's a struct `capturingRequestBidder` in this file that apparently isn't being used anywhere but defines `requestBid`. Can you add the `executor hookexecution.StageExecutor` parameter to that `requestBid` signature please?

File: hooks/hookexecution/executor_test.go [hunk 2]
```diff
@@ -677,6 +678,241 @@ func TestExecuteRawAuctionStage(t *testing.T) {
...
+									Warnings: nil,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			description:            "Stage execution can be timed out",
```
Comment (bsardo) [inline hooks/hookexecution/executor_test.go]: Nitpick: same comment as the other PR. I think a better description would be "Response can be changed when a hook times out"

File: hooks/hookexecution/executor_test.go [hunk 3]
```diff
@@ -677,6 +678,241 @@ func TestExecuteRawAuctionStage(t *testing.T) {
...
+			expectedReject:         &RejectError{0, HookID{ModuleCode: "foobar", HookCode: "foo"}, hooks.StageRawBidderResponse.String()},
+			expectedModuleContexts: foobarModuleCtx,
+			expectedStageOutcomes: []StageOutcome{
+				{
+					ExecutionTime: ExecutionTime{},
+					Entity:        vEntity,
+					Stage:         hooks.StageRawBidderResponse.String(),
+					Groups: []GroupOutcome{
+						{
+							ExecutionTime: ExecutionTime{},
```
Comment (bsardo) [inline hooks/hookexecution/executor_test.go]: Same comment as in the other PR: I think we can remove all of these `ExecutionTime` fields with zero-value values.

**Review by bsardo — COMMENTED** [review]
Looks good. Just a couple of minor comments.

File: exchange/exchange_test.go (hunk 1 above)
Comment (legendko) [reply exchange/exchange_test.go]: I went a bit further and changed the alternateBidderCodes type parameter to be a pointer so the capturingRequestBidder actually implements the AdaptedBidder interface

**Review by bsardo — APPROVED** [review]

**Comment by legendko** [comment]
merged

**Pull request closed without merging**